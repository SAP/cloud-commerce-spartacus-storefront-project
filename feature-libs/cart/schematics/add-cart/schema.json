{
  "$schema": "http://json-schema.org/schema",
  "id": "CartSchematics",
  "title": "Cart Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the Saved cart feature.",
      "default": true
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "items": {
        "enum": ["Saved-cart"],
        "type": "string"
      },
      "default": ["Saved-cart"],
      "x-prompt": "Which features would you like to setup?"
    }
  },
  "required": []
}
