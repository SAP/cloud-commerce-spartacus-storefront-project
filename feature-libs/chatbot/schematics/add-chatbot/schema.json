{
  "$schema": "http://json-schema.org/schema",
  "$id": "ChatBotSchematics",
  "title": "ChatBot Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the chat bot features.",
      "default": true
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "default": ["ChatBot"]
    }
  },
  "required": []
}
