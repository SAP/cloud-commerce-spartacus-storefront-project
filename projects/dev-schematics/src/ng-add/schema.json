{
  "$schema": "http://json-schema.org/schema",
  "id": "SpartacusDevSchematics",
  "title": "Spartacus Dev Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "default-base-sites": {
      "type": "boolean",
      "description": "Add default base site list to the project.",
      "default": false,
      "aliases": ["baseSites"]
    },
    "default-routing": {
      "type": "boolean",
      "description": "Add default routing to the project.",
      "default": false,
      "aliases": ["routing"]
    },
    "test-outlets": {
      "type": "boolean",
      "description": "Add test outlets to the project.",
      "default": false,
      "aliases": ["outlets"]
    }
  },
  "required": []
}
