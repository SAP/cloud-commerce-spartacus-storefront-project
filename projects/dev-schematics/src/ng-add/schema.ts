export interface Schema {
  project: string;
  'default-base-sites': boolean;
  'default-routing': boolean;
  'test-outlets': boolean;
}
