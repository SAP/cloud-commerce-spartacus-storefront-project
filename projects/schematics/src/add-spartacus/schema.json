{
  "$schema": "http://json-schema.org/schema",
  "id": "SpartacusSchematics",
  "title": "Spartacus Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "baseUrl": {
      "type": "string",
      "description": "The url of the OCC backend",
      "default": "https://localhost:9002"
    },
    "baseSite": {
      "type": "string",
      "description": "The base site to use with spartacus",
      "default": "electronics"
    },
    "useMetaTags": {
      "type": "boolean",
      "description": "Configure baseUrl via meta tag in index.html"
    },
    "featureLevel": {
      "type": "string",
      "description": "Configure feature level for application.",
      "default": "1.1"
    },
    "overwriteAppComponent": {
      "type": "boolean",
      "description": "Overwrite whole app.component.html file.",
      "default": false
    }
  },
  "required": []
}
