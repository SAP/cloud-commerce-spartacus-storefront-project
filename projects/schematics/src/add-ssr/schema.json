{
  "$schema": "http://json-schema.org/schema",
  "id": "SpartacusSchematics",
  "title": "Spartacus Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    }
  },
  "required": []
}
