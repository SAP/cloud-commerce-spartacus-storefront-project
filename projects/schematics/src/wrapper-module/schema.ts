import { ExecutionOptions } from '@angular-devkit/schematics';

export interface Schema extends Partial<ExecutionOptions> {
  /**
   * The name of the project in which to execute schematics.
   */
  project: string;
  /**
   * When enabled, prints the additional logs.
   */
  debug?: boolean;
  /**
   * The marker module for which
   * to find or generate a new wrapper module.
   */
  markerModuleName: string;
  /**
   * Name of the feature module
   * to add to the wrapper module.
   */
  featureModuleName: string;
  /**
   * Internal options.
   * Should not be set by the user.
   */
  internal?: {
    /**
     * The execution sequence.
     * Used for ensuring one log message is printed.
     */
    sequence?: number;
  };
}
