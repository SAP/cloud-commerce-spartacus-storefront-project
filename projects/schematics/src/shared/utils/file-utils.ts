import { experimental, strings } from '@angular-devkit/core';
import { SchematicsException, Tree } from '@angular-devkit/schematics';
import { getProjectTargetOptions } from '@angular/cdk/schematics';
import { parseTsconfigFile } from '@angular/core/schematics/utils/typescript/parse_tsconfig';
import { getSourceNodes } from '@schematics/angular/utility/ast-utils';
import {
  Change,
  InsertChange,
  ReplaceChange,
} from '@schematics/angular/utility/change';
import { dirname, relative } from 'path';
import * as ts from 'typescript';

export function getTsSourceFile(tree: Tree, path: string): ts.SourceFile {
  const buffer = tree.read(path);
  if (!buffer) {
    throw new SchematicsException(`Could not read file (${path}).`);
  }
  const content = buffer.toString();
  const source = ts.createSourceFile(
    path,
    content,
    ts.ScriptTarget.Latest,
    true
  );

  return source;
}

export function getAllTsSourceFiles(
  tsconfigPath: string,
  tree: Tree,
  basePath: string
): ts.SourceFile[] {
  const parsed = parseTsconfigFile(tsconfigPath, dirname(tsconfigPath));
  const host = createMigrationCompilerHost(tree, parsed.options, basePath);
  const program = ts.createProgram(parsed.fileNames, parsed.options, host);
  return program
    .getSourceFiles()
    .filter(
      f => !f.isDeclarationFile && !program.isSourceFileFromExternalLibrary(f)
    );
}

export function getIndexHtmlPath(
  project: experimental.workspace.WorkspaceProject
): string {
  const buildOptions = getProjectTargetOptions(project, 'build');

  if (!buildOptions.index) {
    throw new SchematicsException('"index.html" file not found.');
  }

  return buildOptions.index;
}

export function getPathResultsForFile(
  tree: Tree,
  file: string,
  directory?: string
): string[] {
  const results: string[] = [];
  const dir = directory || '/';

  tree.getDir(dir).visit(filePath => {
    if (filePath.endsWith(file)) {
      results.push(filePath);
    }
  });

  return results;
}

export enum InsertDirection {
  LEFT,
  RIGHT,
}

export function commitChanges(
  host: Tree,
  path: string,
  changes: Change[] | null,
  insertDirection: InsertDirection
): void {
  if (!changes) {
    return;
  }

  const recorder = host.beginUpdate(path);
  changes.forEach(change => {
    if (change instanceof InsertChange) {
      const pos = change.pos;
      const toAdd = change.toAdd;
      if (insertDirection === InsertDirection.LEFT) {
        recorder.insertLeft(pos, toAdd);
      } else {
        recorder.insertRight(pos, toAdd);
      }
    } else if (change instanceof ReplaceChange) {
      const pos = change['pos'];
      const oldText = change['oldText'];
      const newText = change['newText'];

      recorder.remove(pos, oldText.length);
      if (insertDirection === InsertDirection.LEFT) {
        recorder.insertLeft(pos, newText);
      } else {
        recorder.insertRight(pos, newText);
      }
    }
  });
  host.commitUpdate(recorder);
}

export function defineProperty(
  nodes: ts.Node[],
  path: string,
  toAdd: string
): InsertChange {
  const constructorNode = nodes.find(n => n.kind === ts.SyntaxKind.Constructor);

  if (!constructorNode) {
    throw new SchematicsException(`No constructor found in ${path}.`);
  }

  return new InsertChange(path, constructorNode.pos + 1, toAdd);
}

export function injectService(
  nodes: ts.Node[],
  path: string,
  serviceName: string,
  propertyName?: string
): InsertChange {
  const constructorNode = nodes.find(n => n.kind === ts.SyntaxKind.Constructor);

  if (!constructorNode) {
    throw new SchematicsException(`No constructor found in ${path}.`);
  }

  const parameterListNode = constructorNode
    .getChildren()
    .find(n => n.kind === ts.SyntaxKind.SyntaxList);

  if (!parameterListNode) {
    throw new SchematicsException(
      `No no parameter list found in ${path}'s constructor.`
    );
  }

  propertyName = propertyName
    ? strings.camelize(propertyName)
    : strings.camelize(serviceName);
  const toAdd = `private ${propertyName}: ${strings.classify(serviceName)}`;
  return new InsertChange(path, parameterListNode.pos, toAdd);
}

export function insertCommentAboveIdentifier(
  sourcePath: string,
  source: ts.SourceFile,
  identifierName: string,
  comment: string
): InsertChange[] {
  const callExpressionNodes = findNodesByTextAndKind(
    source,
    identifierName,
    ts.SyntaxKind.Identifier
  );
  const changes: InsertChange[] = [];
  callExpressionNodes.forEach(n =>
    changes.push(
      new InsertChange(
        sourcePath,
        getLineStartFromTSFile(source, n.getFullStart()),
        comment
      )
    )
  );
  return changes;
}

export function renameIdentifierNode(
  sourcePath: string,
  source: ts.SourceFile,
  oldName: string,
  newName: string
): ReplaceChange[] {
  const callExpressionNodes = findNodesByTextAndKind(
    source,
    oldName,
    ts.SyntaxKind.Identifier
  );
  const changes: ReplaceChange[] = [];
  callExpressionNodes.forEach(n =>
    changes.push(new ReplaceChange(sourcePath, n.getStart(), oldName, newName))
  );
  return changes;
}

function findNodesByTextAndKind(
  source: ts.SourceFile,
  text: string,
  syntaxKind: ts.SyntaxKind
): ts.Node[] {
  const nodes = getSourceNodes(source);
  return nodes
    .filter(n => n.kind === syntaxKind)
    .filter(n => n.getText() === text);
}

function getLineStartFromTSFile(
  source: ts.SourceFile,
  position: number
): number {
  const lac = source.getLineAndCharacterOfPosition(position);
  const lineStart = source.getPositionOfLineAndCharacter(lac.line, 0);

  return lineStart;
}

// as this is copied from https://github.com/angular/angular-cli/blob/master/packages/schematics/angular/app-shell/index.ts#L211, no need to test Angular's code
export function getMetadataProperty(
  metadata: ts.Node,
  propertyName: string
): ts.PropertyAssignment {
  const properties = (metadata as ts.ObjectLiteralExpression).properties;
  const property = properties
    .filter(prop => prop.kind === ts.SyntaxKind.PropertyAssignment)
    .filter((prop: ts.PropertyAssignment) => {
      const name = prop.name;
      switch (name.kind) {
        case ts.SyntaxKind.Identifier:
          return (name as ts.Identifier).getText() === propertyName;
        case ts.SyntaxKind.StringLiteral:
          return (name as ts.StringLiteral).text === propertyName;
      }

      return false;
    })[0];

  return property as ts.PropertyAssignment;
}

// copied from https://github.com/angular/angular/blob/master/packages/core/schematics/utils/typescript/compiler_host.ts#L12, no need to test angular's code
export function createMigrationCompilerHost(
  tree: Tree,
  options: ts.CompilerOptions,
  basePath: string,
  fakeRead?: (fileName: string) => string | null
): ts.CompilerHost {
  const host = ts.createCompilerHost(options, true);

  // We need to overwrite the host "readFile" method, as we want the TypeScript
  // program to be based on the file contents in the virtual file tree. Otherwise
  // if we run multiple migrations we might have intersecting changes and
  // source files.
  host.readFile = fileName => {
    const treeRelativePath = relative(basePath, fileName);
    const fakeOutput = fakeRead ? fakeRead(treeRelativePath) : null;
    const buffer =
      fakeOutput === null ? tree.read(treeRelativePath) : fakeOutput;
    // Strip BOM as otherwise TSC methods (Ex: getWidth) will return an offset,
    // which breaks the CLI UpdateRecorder.
    // See: https://github.com/angular/angular/pull/30719
    return buffer ? buffer.toString().replace(/^\uFEFF/, '') : undefined;
  };

  return host;
}
