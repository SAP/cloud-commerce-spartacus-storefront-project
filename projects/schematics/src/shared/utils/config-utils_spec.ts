/// <reference types="jest" />

import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import * as path from 'path';
import { InMemoryFileSystemHost, Project, SourceFile } from 'ts-morph';
import ts from 'typescript';
import {
  createNewConfig,
  getConfig,
  getConfigs,
  getExistingStorefrontConfigNode,
  getSpartacusProviders,
  mergeConfig,
} from './config-utils';
import { commitChanges, getTsSourceFile } from './file-utils';

const collectionPath = path.join(__dirname, '../../collection.json');
const schematicRunner = new SchematicTestRunner('schematics', collectionPath);

// TODO:#10744 - cleanup after implementing the new config utils.
describe('Storefront config utils', () => {
  let appTree: UnitTestTree;
  const workspaceOptions: WorkspaceOptions = {
    name: 'workspace',
    version: '0.5.0',
  };
  const appOptions: ApplicationOptions = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: false,
    projectRoot: '',
  };
  const defaultOptions = {
    project: 'schematics-test',
  };
  const appModulePath = 'src/app/app.module.ts';

  beforeEach(async () => {
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
    appTree = await schematicRunner
      .runSchematicAsync('add-spartacus', defaultOptions, appTree)
      .toPromise();
  });

  describe.skip('getExistingStorefrontConfigNode', () => {
    it('should get the Storefront config from app.module.ts file', async () => {
      const appModuleFile = getTsSourceFile(appTree, appModulePath);
      const config = getExistingStorefrontConfigNode(
        appModuleFile
      ) as ts.CallExpression;

      expect(config).toBeTruthy();
      expect(config.getFullText()).toContain('B2cStorefrontModule.withConfig');
    });
  });

  describe.skip('getConfig', () => {
    it('should return the specified config from Storefront CallExpression AST node object', async () => {
      const appModuleFile = getTsSourceFile(appTree, appModulePath);
      const config = getExistingStorefrontConfigNode(
        appModuleFile
      ) as ts.CallExpression;
      const currentContextConfig = getConfig(config, 'context');

      expect(currentContextConfig).toBeTruthy();
      expect(currentContextConfig?.getFullText()).toContain('currency:');
    });

    it('should return an undefined if the provided configName was not found', async () => {
      const appModuleFile = getTsSourceFile(appTree, appModulePath);
      const config = getExistingStorefrontConfigNode(
        appModuleFile
      ) as ts.CallExpression;
      const configByName = getConfig(config, 'test');

      expect(configByName).toBeFalsy();
      expect(configByName).toEqual(undefined);
    });
  });

  describe.skip('mergeConfig', () => {
    it('should merge the provided config array', async () => {
      const appModuleFile = getTsSourceFile(appTree, appModulePath);
      const config = getExistingStorefrontConfigNode(
        appModuleFile
      ) as ts.CallExpression;
      const currentContextConfig = getConfig(
        config,
        'context'
      ) as ts.PropertyAssignment;
      const currencyChange = mergeConfig(
        appModulePath,
        currentContextConfig,
        'currency',
        ['EUR', 'JPY']
      );

      expect(appTree.readContent(appModulePath)).not.toContain('EUR');
      expect(appTree.readContent(appModulePath)).not.toContain('JPY');

      commitChanges(appTree, appModulePath, [currencyChange]);

      expect(appTree.readContent(appModulePath)).toContain('EUR');
      expect(appTree.readContent(appModulePath)).toContain('JPY');
    });

    it('should merge the provided regular config', async () => {
      const appModuleFile = getTsSourceFile(appTree, appModulePath);
      const config = getExistingStorefrontConfigNode(
        appModuleFile
      ) as ts.CallExpression;
      const backendConfig = getConfig(
        config,
        'backend'
      ) as ts.PropertyAssignment;

      const change = mergeConfig(appModulePath, backendConfig, 'occ', 'random');

      expect(appTree.readContent(appModulePath)).not.toContain('random');
      commitChanges(appTree, appModulePath, [change]);
      expect(appTree.readContent(appModulePath)).toContain('random');
    });

    it('should create a new config if there is nothing to be mergex', async () => {
      const appModuleFile = getTsSourceFile(appTree, appModulePath);
      const config = getExistingStorefrontConfigNode(
        appModuleFile
      ) as ts.CallExpression;
      const currentContextConfig = getConfig(
        config,
        'context'
      ) as ts.PropertyAssignment;
      const baseSiteChange = mergeConfig(
        appModulePath,
        currentContextConfig,
        'urlParameters',
        ['baseSite', 'language', 'currency']
      );

      expect(appTree.readContent(appModulePath)).not.toContain(
        'urlParameters:'
      );

      commitChanges(appTree, appModulePath, [baseSiteChange]);

      expect(appTree.readContent(appModulePath)).toContain('urlParameters:');
    });
  });

  describe.skip('createNewConfig', () => {
    it('should nest the given new config in the given config object', async () => {
      const appModuleFile = getTsSourceFile(appTree, appModulePath);
      const config = getExistingStorefrontConfigNode(
        appModuleFile
      ) as ts.CallExpression;
      const currentContextConfig = getConfig(
        config,
        'context'
      ) as ts.PropertyAssignment;
      const testConfigChange = createNewConfig(
        appModulePath,
        currentContextConfig,
        'testObjectConfig',
        ['value1', 'value2']
      );

      expect(appTree.readContent(appModulePath)).not.toContain(
        'testObjectConfig:'
      );

      commitChanges(appTree, appModulePath, [testConfigChange]);

      expect(appTree.readContent(appModulePath)).toContain('testObjectConfig:');
      expect(appTree.readContent(appModulePath)).toContain('value1');
      expect(appTree.readContent(appModulePath)).toContain('value2');
    });
  });

  describe('ts-morph config utils', () => {
    const configFileContent = `
import { NgModule } from '@angular/core';
import {
CartAddEntrySuccessEvent,
CartRemoveEntrySuccessEvent,
provideConfig,
} from '@spartacus/core';
import { NavigationEvent, defaultCmsContentProviders } from '@spartacus/storefront';
import { PersonalizationRootModule } from '@spartacus/tracking/personalization/root';
import { AepModule } from '@spartacus/tracking/tms/aep';
import { BaseTmsModule, TmsConfig } from '@spartacus/tracking/tms/core';
import { GtmModule } from '@spartacus/tracking/tms/gtm';
import { someFunction, someSpread } from '@some/package';

@NgModule({
imports: [
BaseTmsModule.forRoot(),
GtmModule,
AepModule,
PersonalizationRootModule,
],
providers: [
someFunction(),
...someSpread,
...defaultCmsContentProviders,
provideConfig(<TmsConfig>{
  tagManager: {
    gtm: {
      events: [NavigationEvent, CartAddEntrySuccessEvent],
    },
    aep: {
      events: [NavigationEvent, CartRemoveEntrySuccessEvent],
    },
  },
}),
provideConfig({
  featureModules: {
    personalization: {
      module: () =>
        import('@spartacus/tracking/personalization').then(
          (m) => m.PersonalizationModule
        ),
    },
  },
}),
],
})
export class TrackingFeatureModule {}
`;
    let sourceFile: SourceFile;

    beforeAll(() => {
      const project = new Project({
        fileSystem: new InMemoryFileSystemHost(),
      });
      sourceFile = project.createSourceFile('test.ts', configFileContent);
    });

    describe('getConfigs', () => {
      it('should return all configs from provideConfigs calls', () => {
        const configs = getConfigs(sourceFile);
        expect(configs.length).toEqual(2);
        expect(configs[0].getText()).toMatchSnapshot();
        expect(configs[1].getText()).toMatchSnapshot();
      });
    });

    describe('getSpartacusProviders', () => {
      it('should return all providers from spartacus in file', () => {
        const providers = getSpartacusProviders(sourceFile);
        expect(providers.length).toEqual(3);
        expect(providers[0].getText()).toMatchSnapshot();
        expect(providers[1].getText()).toMatchSnapshot();
        expect(providers[2].getText()).toMatchSnapshot();
      });
    });
  });
});
