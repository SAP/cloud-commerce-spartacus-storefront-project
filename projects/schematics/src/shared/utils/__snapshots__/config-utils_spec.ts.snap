// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storefront config utils getConfigs should return all configs from provideConfigs calls 1`] = `
"{
      tagManager: {
        gtm: {
          events: [NavigationEvent, CartAddEntrySuccessEvent],
        },
        aep: {
          events: [NavigationEvent, CartRemoveEntrySuccessEvent],
        },
      },
    }"
`;

exports[`Storefront config utils getConfigs should return all configs from provideConfigs calls 2`] = `
"{
      featureModules: {
        personalization: {
          module: () =>
            import('@spartacus/tracking/personalization').then(
              (m) => m.PersonalizationModule
            ),
        },
      },
    }"
`;
